cpp/src/arrow/status.h

```diff
@@ -286,21 +313,31 @@ class ARROW_EXPORT Status {
   /// text or POSIX code information.
   std::string CodeAsString() const;
 
-  /// \brief Return the StatusCode value attached to this status.
-  StatusCode code() const { return ok() ? StatusCode::OK : state_->code; }
+  /// \brief Return the status code enum value attached to this status.
+  template <typename T = StatusCode>
+  T code(T* = nullptr) const {
+    return static_cast<T>(ok() ? 0 : state_->code);
+  }
+
+  const StatusRange* status_range() const;
 
   /// \brief Return the specific error message attached to this status.
   std::string message() const { return ok() ? "" : state_->msg; }
 
+  static void RegisterStatusRange(const StatusRange* status_range);
+
  private:
   struct State {
-    StatusCode code;
+    uint64_t range_id;
+    int32_t code;
     std::string msg;
   };
   // OK status has a `NULL` state_.  Otherwise, `state_` points to
   // a `State` structure containing the error code and message(s)
   State* state_;
 
+  static std::unordered_map<uint64_t, const StatusRange*> ranges_;
```

Comment:
Seems like we're needing more and more a global `arrow::init()`.

Reply:
Are you afraid of the static initialization here?

Reply:
Yes, seems likely that we'll need more and more controlled static initialization:

- Utf8 utils make use of call_once
- This patch (which other static initializers might depend in proper order, but controllable since we're referencing with int32/int64)
- Gandiva make use of call_once, I think this one can be tricky due to possible failures and the time it takes to init, e.g. it's friendlier to the user if he has visibility (time take, failure, ...)